encodef/encodef.1
.\" Man page for "encodef" and "decodef".
.TH encodef 1 2010-05-10 Wheeler "User commands"
.\"
.\" Manual and code released under the "MIT/X11" license.
.\" Copyright (c) 2010 David A. Wheeler
.\"
.\" Permission is hereby granted, free of charge, to any person obtaining a copy
.\" of this software and associated documentation files (the "Software"), to
.\" deal in the Software without restriction, including without limitation
.\" the rights to use, copy, modify, merge, publish, distribute, sublicense,
.\" and/or sell copies of the Software, and to permit persons to whom the
.\" Software is furnished to do so, subject to the following conditions:
.\"
.\" The above copyright notice and this permission notice shall be included in
.\" all copies or substantial portions of the Software.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
.\" IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
.\" FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
.\" AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
.\" LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
.\" FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
.\" DEALINGS IN THE SOFTWARE.

.SH NAME
encodef \- encode/decode filenames or similar data

.SH SYNOPSIS
.PP
\fBencodef\fR [\fIoptions\fR...] [ \fB--\fR [\fIfilenames...\fR] ]
.PP
\fBdecodef\fR [\fIoptions\fR...] [ \fB--\fR [\fIfilenames...\fR] ]

.SH DESCRIPTION
.PP
POSIX/Unix/Linux filenames and pathnames
can include nearly-arbitrary sequences of bytes (octets),
including newlines, tabs, terminal control sequences, spaces,
sequences that are not legal in the current locale's encoding, and so on.
Many text-processing utilities cannot easily process arbitrary filenames
directly, e.g., they may misinterpret a newline in the middle of a filename
as the end of the filename.
Some tools can generate or accept filenames terminated by a null byte,
but many text-processing utilities cannot easily process these either.
.PP
The \fIencodef\fR and \fIdecodef\fR
utilities make it much easier to process arbitrary filenames.
The \fIencodef\fR utility accepts filenames and sends to stdout
encoded filenames that are much easier to process.
The \fIdecodef\fR utility reverses this; it
accepts encoded filenames and sends to stdout the decoded filenames
that can be directly used.
These utilities can be used for arbitrary data, not just filenames,
but they are specifically designed to work well with filenames.
.PP
More specifically, the \fIencodef\fR
utility accepts 0 or more filenames from its command line
if an "--" end-of-options argument is provided;
or it reads from stdin instead if no "--" is provided.
It sends to stdout the encoding of each filename, in order,
with each encoded filename terminated by a newline.
If stdin is read, the filenames are separated by a null byte
(newline cannot be used since filenames can include newlines).
The last filename from stdin may, but need not, end with a null byte.
This processing of stdin
enables easy processing or transition from filename lists where
filenames are terminated or separated by a null byte.
By default, \fIencodef\fR uses percent encoding (e.g., %hh),
though other encodings are available.
.PP
When encoding, the tool must determine which bytes (characters)
are encoded.
By default, \fIencodef\fR encodes most characters except
A-Z, a-z, 0-9, <slash> (/), <period> (.), <underscore> (_),
and non-leading <hyphen> (-).
A dash is leading if it is at the beginning of a name or immediately
follows a forward slash.
This encodes most characters, and is the default because it results
in the safest possible filenames, reducing the risk of errors or
security vulnerabilities due to mishandled filenames.
In particular, this default escapes:
(1) the control characters including newline, tab, and terminal
control characters,
(2) the space character, so shell users can use the default IFS setting,
and
(3) the shell globbing characters * and ?.
In addition, by default all of the unescaped characters
and the escape characters are in the POSIX
"portable character set", so they \fImust\fR exist in any given locale.
(This also means that all other letters are encoded, but this is
often a good thing;
there is no way to be certain of a filename's character encoding
in the general case.)
Various options (see -C and -S) control which characters are encoded.
.PP
The decodef utility reverses \fIencodef\fR;
the encodef -d (decode) option makes encodef perform the work of decodef.
It accepts encoded filenames from the command line
(or from stdin if no "--" option is provided) and decodes the filenames
back to their original form, sending the result to stdout.
If only a single filename is provided,
the output on stdout is simply the undecoded name.
When reading from stdin, each filename is terminated by newline or null byte;
the last filename need not be terminated.
By default, if multiple filenames are provided, the output on stdout is the
undecoded names separated by a null byte
(since this is the only value that cannot occur in a filename;
the last value is not terminated with a null byte).
One odd option, -Y, appends the letter Y to the end of each filename,
and is needed to properly decode filenames using
shell command substitution (as explained below).
.PP
Several different filename encodings are supported:
.IP \(bu 4
-U: URI encoding, aka URL encoding aka
percent encoding. In this encoding, the percent
character % is the escape
character, and %hh represents
the byte with the hexadecimal value hh (in either upper or lower case).
Either %% or %25 is accepted as an encoding of the percent character %,
but %25 is what is generated when encoding %.
This encoding is described and used in IETF RFC 3986.
This particular option flag was selected because -U is not used by
GNU xargs, and %U is not used in GNU printf(1);
GNU is both common and very feature-rich, so it seemed likely that these
flag values wouldn't be used by others.
Unfortunately, there don't seem to be many flags that aren't already
used by find's -printf;
GNU find -printf already uses %p, %P, %u, and %U for other things.
.IP \(bu 4
-B: traditional backslash encoding.
In this encoding, the backslash character \\ is the escape character.
The expression \\ddd, where ddd is a 1 to 3 digit octal value,
can represent an arbitrary byte;
note that a leading 0 is not required, so byte 255 is represented
as \\377, and not \\0377 as with printf(1)'s %b.
The usual escape sequences are also accepted from POSIX: '\\\\' (for backslash itself), '\\a', '\\b', '\\f', '\\n', '\\r', '\\t', and '\\v'.
When decoding, a backslash followed a non-alphanumeric character
is replaced by that character, so \\" encodes ".
As an extension, this decoder also accepts \\xHH as a representation of the
hexadecimal number HH in upper or lower case
(the encoder does not generate this, but some other programs accept this
representation; localedef(1) supports this notation, for example).
This encoding for backslash
is used by printf(1) format strings, printf(3), C, C++, tr(1),
and many other tools.
However, beware of providing encoded filenames as the format string values to
most such tools directly, because they will often interpret other
characters such as '%' in a way that will cause them to fail (and in
some cases can lead to a security vulnerability).
The \fIdecode\fR utility does \fInot\fR interpret other characters like '%'
specially, and thus does not have this risk.
This is \fInot\fR the encoding used by printf %b (aka "pfb" or "extra 0"
encoding).
This particular option flag was selected because -B is not used by
GNU xargs, and %B is not used in GNU printf(1) (%b is part of the
POSIX standard, but that's different).
GNU find -printf does not use %B, which is nice (so that COULD be used).
.IP \(bu 4
-b (extra 0): printf(1) %b (pfb) encoding.
This is the format decoded by POSIX's printf(1) %b format
(but \fInot\fR by the printf(1) format string itself,
nor the format used by other POSIX tools).
In this encoding, the backslash character \\ is the escape character.
The expression \\0ddd, where ddd is a 0 to 3 digit octal value,
can represent an arbitrary byte;
note that a leading "extra" 0 is always required with octal codes,
so byte 255 is represented as \\0377 and not \\377.
It also accepts and generates the POSIX encodings '\\\\' (for backslash itself), '\\a', '\\b', '\\f', '\\n', '\\r', '\\t', and '\\v'.
Note that printf(1) %b does \fInot\fR portably accept some other
common escape patterns, such as \\" or \\' or \\ followed by a space.
This is an odd encoding, and not what many people think it is.
Its main advantages are that
it is portable (because it is part of POSIX.1-2008 printf(1)),
widely implemented, and
often efficient in shells (because printf(1) is typically a shell built-in).
This particular option flag was selected because -b is not used by
GNU xargs, and %b is the POSIX-required name for the relevant printf(1).
Unfortunately,
GNU find -printf already uses %b for something else.
(Older versions of this tool use -e for "extra 0".)
.PP
Note that for purposes of these utilities, encodings have little
to do with the character encoding used for internationalization.
In POSIX, filenames are simply sequences of bytes;
there is no standard mechanism to determine what character encoding
was used to encode a given filename.
The author recommends that users always use UTF-8 as the character
encoding to encode all filenames.

.SH OPTIONS

.PP
Options can control factors such as
which encoding scheme is used (see -B, -b, -U),
which bytes will be encoded (see -C and -S),
and the format of the results (see -Y).
The following options are supported:

.TP
\fB\-B\fR
Use traditional backslash encoding.
The expression \\ddd, where ddd is a 1 to 3 digit octal value,
represent an arbitrary byte.
The encoder produces, and the decoder accepts, the usual
expressions '\\\\' (for backslash itself), '\\a', '\\b', '\\f', '\\n', '\\r', '\\t', and '\\v'.
The decoder also accepts '\\"' (for double quote), '\\'' (for single quote), and '\\ ' (for space).
As an extension, this decoder also accepts \\xHH as a representation of the
hexadecimal number HH in upper or lower case
(the encoder does not generate this, but some other programs accept this
representation).
When encoding, this is the encoding of the \fIoutput\fR;
when decoding, this is the encoding of the \fIinput\fR.
.TP
\fB\-d\fR
Decode.  If passed as an option to encodef, makes it work as decodef.
.TP
\fB\-b\fR
Use the "extra 0" (aka printf %b or pfb) encoding.
This is the encoding accepted by the printf %b format specifier,
though it is not the same as many other POSIX tools.
The expression \\0ddd, where ddd is a 0 to 3 digit octal value,
can represent an arbitrary byte;
note that a leading "extra" 0 is always required with octal codes.
It also accepts and generates the POSIX encodings '\\\\' (for backslash itself), '\\a', '\\b', '\\f', '\\n', '\\r', '\\t', and '\\v'.
Note that printf %b is not guaranteed to accept \\" and other escape
sequences.
When encoding, this is the encoding of the \fIoutput\fR;
when decoding, this is the encoding of the \fIinput\fR.

.TP
\fB\-C\fR
(Encoder only) Only encode control characters and the "escape"
character (\\ or %).
It also encodes "-" if it is at the beginning of a pathname component,
to reduce the risk from leading "-"
(these are often confused with option flags).
See also -S.

.TP
\fB\-S\fR
(Encoder only) Like -C, but also encode the byte representing the space
character.
This is useful when handing filenames to the shell;
the default IFS values (space, tab, newline) are all encoded.
See also -C.

.TP
\fB\-U\fR
Use percent encoding (aka URL encoding), where %HH represents
the hexadecimal value HH.
When encoding, this is the encoding of the \fIoutput\fR;
when decoding, this is the encoding of the \fIinput\fR.

.TP
\fB-Y\fR
(Decoder only) Y-append mode.
Append the letter 'Y' after the end of each filename in the output.
This unusual mode supports shell scripts.
Shell command substitution strips off trailing newlines, which could
corrupt filenames that end with newline.
If a shell script uses command substitution to invoke the decoder directly,
it should use this option to ensure that the last character is
always 'Y' and then strip off the trailing 'Y'.
("X" is not used as the trailing letter; it's being reserved
for options involving xargs and XML.)


.SH "EXIT STATUS"
.SH ENVIRONMENT
.SH FILES
.SH "CONFORMING TO"
.PP
Percent encoding is defined in IETF RFC 3986.
POSIX.1-2008 defines both of the
backslash quotation systems supported by printf(1);
in particular,
POSIX.1-2008 XBD Chapter 5 table 5-1, page 121
defines '\\\\' (for backslash itself), '\\a', '\\b', '\\f', '\\n', '\\r', '\\t', and '\\v'.
.SH NOTES
.PP
The encoder and decoder can process arbitrary filename lengths,
but the underlying filesystem, interfaces, and utilities
will have some sort of limit.

.SH BUGS
.PP
The fact that POSIX systems permit filenames to contain
control characters like newline,
return, tab, and ESC (used for terminal escapes) 
could be considered a bug all by itself;
forbidding them would reduce the need for this pair of tools.
If POSIX systems could never encounter bytes 1 through 31 in filenames,
filename processing would be \fImuch\fR simpler.
For example, once a shell script sets IFS to newline and tab
(e.g., near its beginning):
.PP
 IFS="`printf '\\n\\t'`"
.PP
It could loop over all filenames in the current tree with
the much simpler expression:
.PP
 for f in `find .` ; do ...
.PP
POSIX does not require that filenames meet a particular
character encoding such as UTF-8; we recommend that users always
use UTF-8, to simplify worldwide exchange of data.
.PP
Currently this doesn't have options to control which bytes
will be encoded; I'm sure that will change.

.SH EXAMPLE
.PP
You can display the filenames from the current directory down,
encoded in percent encoding, with:
 find . -exec encodef -p -- {} \\+

.PP
You can loop over arbitrary trees of files in shell with encodef.
Here is one example, which uses -e encoding (which is supported by the
typically built-in printf(1)).
Since newline, tab, and space are all encoded by default, you can
use the default value of IFS and still have correct filenames.
Note the "trailing Y" trick in the second line; filenames can end in newline,
and command substitution strips off trailing newlines, so you need to
something like the following if filenames might end with a newline:
 for ef in `find . -exec encodef -e -- {} \\+` ; do
   filename="$(printf "%bY" "$ef")" ; filename="${filename%Y}" ...

.PP
Here is a similar loop, but using the percent encoding, and
showing how to use decodef.  Note that when decoding filenames
via shell command substitution, you should normally use -Y or
filenames with trailing newlines will be corrupted:
 for ef in `find . -exec encodef -p -- {} \\+` ; do
   filename="$(decodef -pY -- "$ef")" ; filename="${filename%Y}" ...

.PP
You can use the "xargsf" utility to simultaneously decode the filename
and run a program using the filename; this is sometimes easier.
The "xargsf" utility is simply a variant of xargs that
decodes the filenames provided (per decodef),
and then runs the indicated program(s) on those filenames.
(In the future, it's hoped that xargs would add the basic features of
decodef, but having a separate "xargsf" lets people easily experiment.)
Here is an example:
 for f in `find . -exec encodef -p -- {} \\+` ; do
   printf "%s" "$f" | xargsf -p ls -l

.\"  * If it is called with 0 arguments, it operates as a filter that
.\"  * converts null-byte-terminated lists to pfb format. In this case,
.\"  * an embedded newline is output as \0.  This means you can do this:
.\"  *    for ef in `find . -print0 | encodef` ; do
.\"  *      filename="$(printf "%bX" "$ef")" ; filename="${filename%X}" ...
.SH AUTHOR
.PP
Written by David A. Wheeler.

.SH "SEE ALSO"
.PP
printf(1), iconv(1), sh(1), find(1), xargs(1).
.PP
See its web page
http://www.dwheeler.com/encodef



quoter/ChangeLog
2009-08-20 David A. Wheeler <dwheeler, at, dwheeler.com>
	* Released version 1.10
	* Added rules to clean up seamonkey results
	* Made it comply better with some of the GNU source distribution
	  rules, e.g., make supports prefix= and DESTDIR= now.

2004-09-29 David A. Wheeler <dwheeler, at, dwheeler.com>
	* Released version 1.02.
	* Added rule for "<!..." so that <!DOCTYPE...> will
	  be handled correctly.  Be careful - comments begin with <!-- and
	  must be handled differently.



encodef/encodef.c
/* encodef - encode/decode filenames or similar data.
 *
 * USAGE:
 *  encodef [options] [ -- [filenames] ]
 *
 * Encode/decode filenames or other data, with a variety of
 * encodings and options.  This makes it easier to handle filenames
 * with "unusual" bytes like newline or tab, while using ordinary
 * text-processing utilities.
 *
 * Without the "-d" option, it ENCODES data;
 * with the "-d" option, it DECODES data.
 * If "--" is an option, then any following arguments are
 * encoded/decoded in turn, and stdin is ignored.  If there is no "--", then
 * stdin is read, and the resulting encoding/decoding is sent to stdout.
 *
 * This generates a newline-terminated list of values in
 * printf(1) "%b" (pfb) format.  The pfb format uses "\" as the escape
 * character; \n = newline, \t = tab, \r = return, \0ddd = octal value
 * (where ddd is a 0-3 digit value; digits must be unambiguous).
 * This converts all characters except ".", "/", "_", and ASCII alphanumerics
 * into escaped pfb format; this produces "safe" filenames that are
 * easily processed, since control characters (inc. newline and tab),
 * space, dash, and many other troublesome characters are encoded.
 * Escaping uses octal codes except for newline, tab, and return.
 * We encode "\" using the octal escape, and *not* as \\, so that
 * the sequence \<newline> never occurs.
 * WARNING: There are many OTHER formats that use \ as the escape character,
 * and most are subtly incompatible with each other.  Note that in pfb
 * \0001 is control-A, and NOT \0 1.  Note that '\ ' and '\"' are
 * not supported by pfb.
 *
 * If called with 1 or more arguments, the first argument MUST be "--";
 * all arguments after that are printed in pfb format; each encoded
 * argument is terminated by newline.
 * This means you can do this in shell:
 *    for ef in `find . -exec encodef -- {} \+` ; do
 *      filename="$(printf "%bX" "$ef")" ; filename="${filename%X}" ...
 * If it is called with 0 arguments, it operates as a filter that
 * converts null-byte-terminated lists to pfb format. In this case,
 * an embedded newline is output as \0.  This means you can do this:
 *    for ef in `find . -print0 | encodef` ; do
 *      filename="$(printf "%bX" "$ef")" ; filename="${filename%X}" ...
 * This can be useful with filenames; see
 *    http://www.dwheeler.com/essays/filenames-in-shell.html
 *
 * Possible future extensions:
 *  - "-d" for "decode"?  or create a separate decodef program?
 *
 *  WHEN ENCODING - OUTPUT:
 *  - pfb -b
 *  - support %xx escaping (-p for "percent encoding"?)
 *  - support XML escaping, e.g., &lt; (-X for "XML encoding"?)
 *  - xargs escaping instead (ugh!) (-x for xargs?)
 *  - support C-string-style escaping instead. (-c), e.g., \5 and \".
 *  - null byte escaping (useful if given arguments) (-0)
 *  - Generate \0 when given null byte; only useful in filter mode (-z).
 *  - when encoding, encode escape char as duplicate (e.g., %% for %) (-2?)
 *
 *  WHEN DECODING - the previous ones describe the *input*.  Also:
 *  - Append 'Y' to end of output, to work with shell (shell strips trailing
 *    newlines; adding 'Y' to the end means you can get it, and then strip
 *    the ending 'Y', so you don't lose data). -Y?
 *  - Escape so it's suitable for xargs
 *  - If multiple filenames, by default separate with \0.
 *  - Control if last output terminates with \0 or not (default no).
 *  - Control if use \n instead of \0 output between/end of filenames.
 *
 *  WHEN ENCODING - WHAT IS ENCODED:
 *  - permit 128..255 to go through unescaped (-h for "high bit"?)
 *  - Don't escape characters in 33..126 range (-r for "reduced encoding?")
 *    It still has to encode the escape character.
 *  - The list is a list of filenames to read and go through a filter (-l)
 *  - The following is a filename containing files to be filtered (-f)
 *  - Could say "don't encode high-bit-chars if legal UTF-8", but that
 *    would require changing how the code works because multiple bytes
 *    must be analyzed to determine that. (-U?)
 * I expect to always *REQUIRE* a "--" before non-option arguments, because
 * the primary use is for filenames.  Since filenames can begin with "-",
 * and such filenames might be set by an attacker, requiring "--" makes
 * this program much safer to use.
 *
 * Released under the "MIT/X11" license.
 * Copyright (c) 2010 David A. Wheeler
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <ctype.h>
#include <assert.h>

/* Since we never use setlocale(), we will use the raw data of locale "C".
 * That's just what we want, since the locale settings need not match
 * the data we're processing.  */

/* This code should be correct even if char is signed.
 * It should work fine on any system, even those using EBCDIC. */

/* Option settings */
static int encoding;  /* Encoding system to use */
#define ENCODING_PFB 0         /* "Extra 0" */
#define ENCODING_BACKSLASH 1  /* Traditional backslash: \0400 is "space 0" */
#define ENCODING_PERCENT 10   /* %hh */

static int what_to_encode;
#define ENCODE_LOTS 0
#define ENCODE_CONTROL_SPACE 1
#define ENCODE_CONTROL 2

static bool decode; /* True if we're decoding instead of encoding */
static bool ymode;  /* True if "Ymode" - postfix filenames with Y, so that
                     * filenames ending in \n won't get their name
                     * consumed by sh(1)'s command substitution */


static int curr; /* Current byte to output */
static int prev; /* Previous byte that was already output */


/* Return "true" if current byte "curr" should be escaped */
bool should_escape()
{
  assert(curr != -1);
  if ((curr == '-') && ((prev == -1) || (prev == '/'))) {
    /* Always escape '-' if it's the leading char of a pathname component;
     * it often causes confusion with the option char "-" */
    return true;
  } else if (what_to_encode == ENCODE_CONTROL) {
    return iscntrl(curr);
  } else if (what_to_encode == ENCODE_CONTROL_SPACE) {
    return (curr == ' ') || iscntrl(curr);
  } else {
    /* Default: only leave alphanumerics, "/", ".", "_", "-" alone
     * ("-" is escaped if it's the leading char of a pathname component).
     * By default we escape whitespace, control chars, globbing chars, etc.
     * We'll escape ":" so that it's not misunderstood as a URI.
     * This escapes UTF-8 and other non-alphas, but that is a GOOD thing;
     * there is no way to be certain that a byte is of any particular
     * encoding */
    if (isalnum(curr) || (curr == '/') || (curr == '.') || (curr == '_')
        || (curr == '-'))
      return false;
    else
      return true;
  }
}

/* Send to stdout the encoding of byte "curr" which is being followed
 * by passed-in "next".  If "nullnewline" is true, output char 0 as newline.
 * You must call begin_encoding before using this. */
void encodeb(int next)
{
  bool escapeit;

  if (curr == -1) {  /* We haven't output anything yet; store this byte */
    curr = next;
    return;
  }

  if (encoding == ENCODING_PERCENT) {
    escapeit = should_escape();
    if (curr == '%') escapeit = true; /* MUST escape the escape char */
    /* Now output it, escaped if necessary */
    if (escapeit)
      if (curr == 0)
        putchar('\n');
      else
        printf("%%%.2x", ((unsigned int) curr));
    else
      putchar(curr);
  } else {
    switch (curr) {
      case   0 : putchar('\n'); break;  /* TODO */
      case '\\': fputs("\\\\", stdout); break;
      case '\a': fputs("\\a", stdout); break;
      case '\b': fputs("\\b", stdout); break;
      case '\f': fputs("\\f", stdout); break;
      case '\n': fputs("\\n", stdout); break; /* In EBCDIC, this is >32 */
      case '\r': fputs("\\r", stdout); break;
      case '\t': fputs("\\t", stdout); break;
      case '\v': fputs("\\v", stdout); break;
      default:
        escapeit = should_escape();
        if (curr == '\\') escapeit = true; /* MUST escape the escape char */
  
        /* Now output it, escaped if necessary */
        if (escapeit) {
          putchar('\\'); /* It's escaped, output the escape char */
          if (encoding == ENCODING_PFB) putchar('0'); /* Output "extra 0" */
          if (isdigit(next)) {
            /* A digit follows, so use the unambiguous 3-digit format. */
            printf("%.3o", ((unsigned int) curr));
          } else {
            /* We can use a "shorter form", so we'll do so.
             * Other than the null byte, we'll use at least 2 digits;
             * that way, it cannot be mistaken for a backreference.
             * Since \0 cannot be a backreference, and \0 is the usual
             * format for the null byte, we *will* output \0. Outputting
             * 2+ digits is not necessary, but it seems like a good idea
             * and emulates other tools that do this.
             * Note that localedef does this per
             * POSIX.1-2008 Vol. 1 section 7.3 (3).
             * The POSIX rationale for the Character Set Description File,
             * POSIX.1-2008 Rationale section A.6.4, says:
             * "The octal number notation with no leading zero required
             * was selected to match those of awk and tr and is
             * consistent with that used by localedef.
             * To avoid confusion between an octal constant and
             * the back-references used in localedef source, the octal,
             * hexadecimal, and decimal constants must
             * contain at least two digits." */
            if (curr == 0) { /* ENCODING_PFB already has \0, don't dup */
              if (encoding != ENCODING_PFB) putchar('0');
            } else {
              printf("%.2o", ((unsigned int) curr));
            }
          }
        } else {
          putchar(curr);
        }
        break;
    }
  }
  /* Finished processing curr; consume it and establish new state */
  prev = curr;
  curr = next;
}

void begin_encoding()
{
  curr = prev = -1;
}



int fromhex(int c) {
  switch (c) {
    case '0': return 0;
    case '1': return 1;
    case '2': return 2;
    case '3': return 3;
    case '4': return 4;
    case '5': return 5;
    case '6': return 6;
    case '7': return 7;
    case '8': return 8;
    case '9': return 9;
    case 'a': case 'A': return 10;
    case 'b': case 'B': return 11;
    case 'c': case 'C': return 12;
    case 'd': case 'D': return 13;
    case 'e': case 'E': return 14;
    case 'f': case 'F': return 15;
    default: return 0;
  }
}

/* Return true if octal digit, else return false */
bool isodigit(int c)
{
  switch (c) {
    case '0': case '1': case '2': case '3':
    case '4': case '5': case '6':
      return true;
    default:
      return false;
  }
}

static int store[7]; /* The "store" of submitted values to decode */
static int num_store;

/* Print state of store.  Not used directly; left here for debugging */
void print_store()
{
  int i;
  printf("Store (size=%d) is now:", num_store);
  for (i = 0; i < num_store; i++)
    printf(" %c", store[i]);
  printf("\n");
}

/* Shift (remove) bytes in storage by "amount" */
void shift(int amount)
{
  int i;
  for (i = amount; i < num_store ; i++) {
    store[i - amount] = store[i];
  }
  num_store -= amount;
}

/* Send in a byte c for decoding */
void decodeb(int c)
{
  int i;
  int numleft;
  int currentvalue;
  store[num_store++] = c;

  if (encoding == ENCODING_PERCENT) {
    if (c <= 0) {  /* Stop decoding */
      num_store--;
      if (num_store > 0) {
        /* %hh ended prematurely; output the undecoded characters. */
        putchar(store[0]);
        if (num_store > 1) putchar(store[1]);
      }
      num_store = 0;
    } else if (store[0] != '%') { /* Normal byte, just output it. */
      assert((num_store == 1) && (store[0] == c));
      putchar(c);
      num_store = 0;
    } else if ((num_store == 2) && (c == '%')) {
      /* Decode %% */
      assert((store[0] == '%') && (store[1] == '%'));
      putchar('%');
      num_store = 0;
    } else if (num_store >= 3) {
      assert(store[0] == '%');
      if (isxdigit(store[1]) && isxdigit(store[2])) {
        putchar(fromhex(store[1])*16+fromhex(store[2]));
      } else {
        /* Bad hh after %; just print the bytes. */
        putchar(store[0]);
        putchar(store[1]);
        putchar(store[2]);
      }
      num_store -= 3; /* Consume 3 characters */
    } /* Else, we're still reading in a %xx expression, so do nothing yet */
  } else { /* ENCODING_PFB or ENCODING_BACKSLASH */
    while ((num_store > 0) && (store[0] != '\\') && (store[0] != -1)) {
      /* Handle bytes not beginning with "\" */
      putchar(store[0]);
      shift(1);
    }
    if ((num_store == 1) && (c < 0)) {
      num_store--; /* We're at EOF, do nothing */
    } else if ((c < 0) || (num_store >= 6)) {
      /* A \, followed by enough characters to be unambiguous */
      if (store[0] != '\\') {
        printf("Uh oh, store[0] = %d\n", store[0]);
      }
      assert(store[0] == '\\');
      for (;;) {
        assert(num_store >= 0);
        if (num_store == 0) break;  /* Nothing left to process */
        if (store[0] != '\\') {
          putchar(store[0]);
          shift(1);
          break;
        }
        if (num_store == 1) {  /* One char with nothing following it. */
          putchar(store[0]);
          shift(1);
          break;
        }
        switch (store[1]) {
          case '\\': putchar('\\'); shift(2); break;
          case 'a': putchar('\a'); shift(2); break;
          case 'b': putchar('\b'); shift(2); break;
          case 'f': putchar('\f'); shift(2); break;
          case 'n': putchar('\n'); shift(2); break;
          case 'r': putchar('\r'); shift(2); break;
          case 't': putchar('\t'); shift(2); break;
          case 'v': putchar('\v'); shift(2); break;
          /* NOT HANDLED: \c in printf %b: "produce no further output" */
          /* NOT HANDLED: \hXX */
          case '0': case '1': case '2': case '3':
          case '4': case '5': case '6':
            /* Handle \octaldigits, including PFB's \0octaldigits */
            i = 1; /* i points to the current octal character */
            if ((encoding == ENCODING_PFB) && (store[1] == '0')) {
              /* PFB has a weird extra "0" after \; handle that */
              i++;
              if (!isodigit(store[2])) { /* Must handle specially */
                putchar('\0'); shift(2); break;
              }
            }
            numleft = 3;
            currentvalue = 0;
            /* Handle 1-3 digit octal value. */
            while ((numleft > 0) && (isodigit(store[i]))) {
              currentvalue = currentvalue * 8 + (store[i] - '0');
              i++;
              numleft--;
            }
            putchar(currentvalue);
            shift(i);
            break;
          default:
            putchar(store[1]); shift(2); break; /* Print following byte */
        }
        if (c >= 0) break; /* More characters to read, so don't do more. */
        if (num_store == 0) break; /* Nothing left to process. */
      }
    } /* Otherwise, let's just read the next character in */
  }
  if ((c <= 0) && ymode) putchar('Y');
  /* TODO: Output \0 separators/terminators */
}

void begin_decoding()
{
  num_store = 0;
}


void process_options(const char *options)
{
  const char *p;
  if (options[1] == '-') {
    fprintf(stderr, "No longname options available\n");
    exit(1);
  }
  for (p=options+1; *p; p++)
    switch (*p) {
      case 'd':  decode = true; break;
      case 'U':  encoding = ENCODING_PERCENT; break;
      case 'b':  encoding = ENCODING_PFB; break;
      case 'B':  encoding = ENCODING_BACKSLASH; break;
      case 'Y':  ymode = true; break;
      case 'C':  what_to_encode = ENCODE_CONTROL; break;
      case 'S':  what_to_encode = ENCODE_CONTROL_SPACE; break;
      default:
        fprintf(stderr, "Unknown option %c\n", *p);
        exit(1);
    }
}

int main(int argc, const char* argv[])
{
  int i, c;
  char *p;
  int num_parameters = argc - 1;

  /* Set up defaults */
  encoding = ENCODING_PERCENT; /* Should it be ENCODING_PFB? */
  what_to_encode = ENCODE_LOTS;
  decode = false;
  ymode = false;

  /* Process arguments */
  for (i = 1; i <= num_parameters; i++) {
    if (!strcmp(argv[i], "--")) break; /* Stop processing with -- */
    if (!strcmp(argv[i], "-")) {
      fprintf(stderr, "Single '-' not accepted\n");
      exit(1);
    } else if ((argv[i])[0] != '-') {
      fprintf(stderr, "Must separate filenames from options with '--'\n");
      exit(1);
    } else process_options(argv[i]);
  }

  if (decode) {
    if (i <= num_parameters) {
      if (strcmp(argv[i], "--")) {
        fprintf(stderr, "First non-option must be '--'.\n");
        exit(1);
      }
      /* Process remaining argument list as a set of unescaped filenames */
      for (i++; i <= num_parameters; i++) {
        begin_encoding();
        for (p = (char *) argv[i]; *p != 0; p++)
          decodeb(*p);
        /* Send the terminator */
        decodeb(-1);
      }
    } else {
      /* Process stdin as a set of filenames; filenames are
         terminated by newline or null byte.  */
      begin_encoding();
      while ( (c = getchar()) != EOF) {
        if (c == '\n')  decodeb(0);
        else            decodeb(c);
      }
      decodeb(-1); /* Output the last byte. */
    }
  } else {  /* Encode, not decode */
    if (i <= num_parameters) {
      if (strcmp(argv[i], "--")) {
        fprintf(stderr, "First non-option must be '--'.\n");
        exit(1);
      }
      /* Process remaining argument list as a set of unescaped filenames */
      for (i++; i <= num_parameters; i++) {
        begin_encoding();
        for (p = (char *) argv[i]; *p != 0; p++)
          encodeb(*p);
        /* Send the terminator */
        encodeb(0);
        encodeb(-1);
      }
    } else {
      /* Process stdin as a set of null-byte-terminated values.  Null byte
       * termination is a common for handling "bad" filenames, and this lets
       * people transition from null byte termination to another encoding. */
      begin_encoding();
      while ( (c = getchar()) != EOF)
        encodeb(c);
      encodeb(-1); /* Output the last byte. */
    }
  }
  return 0;
}

